Content.Server/Chemistry/EntitySystems/InjectorSystem.cs

```diff
@@ -330,8 +330,23 @@ private void AfterDraw(Entity<InjectorComponent> injector, EntityUid target)
             return;
         }
 
+        // Frontier - Reagent Whitelist
+        var applicableTargetSolution = targetSolution.Comp.Solution;
+        // If a whitelist exists, remove all non-whitelisted reagents from the target solution temporarily
+        Solution temporarilyRemovedSolution = new();
+        if (injector.Comp.ReagentWhitelist is { } reagentWhitelist)
+        {
+            string[] reagentPrototypeWhitelistArray = new string[reagentWhitelist.Count];
+            for (var i = 0; i < reagentWhitelist.Count; ++i)
+            {
+                reagentPrototypeWhitelistArray[i] = reagentWhitelist[i];
+            }
+            temporarilyRemovedSolution = applicableTargetSolution.SplitSolutionWithout(applicableTargetSolution.Volume, reagentPrototypeWhitelistArray);
```

Comment:
Because reagentWhitelist is a List, isn't lookup linear time?  An iterator would avoid this, no?  Also, why "Prototype"?

```suggestion
            string[] reagentWhitelistArray = new string[reagentWhitelist.Count];
            int i=0;
            foreach (var reagent in reagentWhitelist)
            {
                reagentWhitelistArray[i] = reagent;
            }
            temporarilyRemovedSolution = applicableTargetSolution.SplitSolutionWithout(applicableTargetSolution.Volume, reagentWhitelistArray);
```

Reply:
I am a bit confused, wouldn't the foreach still be linear anyways? Is reagentWhitelist.Count actually a linear big O function?

Reply:
The name Prototype is because the name of the second parameter for SplitSolutionWithout is excludedPrototypes. The term PrototypeID might be better to indicate it's a string instead of an actual prototype.

Reply:
As far as I understand, foreach should be linear (iterator keeps its position), yes, but indexing at [i] is also linear, so the for loop should be quadratic.

Bit of a moot point since n is 1 here, and I haven't benchmarked any of this.

Reply:
Right,, if this has an actual affect on performance, I'm fine with changing it around using a foreach. I used a for loop because I was using a number for indexing.